graph-data.rs/src/gpg.rs

```diff
@@ -0,0 +1,85 @@
+/// Tiny crate to verify message signature and format
+use anyhow::Result as Fallible;
+use anyhow::{format_err, Context};
+use bytes::buf::BufExt;
+use bytes::Bytes;
+use serde::Deserialize;
+use serde_json;
+use std::fs::{read_dir, File};
+
+use pgp::composed::message::Message;
+use pgp::composed::signed_key::SignedPublicKey;
+use pgp::Deserializable;
+
+// Location of public keys
+static PUBKEYS_DIR: &str = "/usr/local/share/public-keys/";
+
+// Signature format
+#[derive(Deserialize)]
+struct SignatureImage {
+  #[serde(rename = "docker-manifest-digest")]
+  digest: String,
+}
+
+#[derive(Deserialize)]
+struct SignatureCritical {
+  image: SignatureImage,
+}
+
+#[derive(Deserialize)]
+struct Signature {
+  critical: SignatureCritical,
+}
+
+/// Keyring is a collection of public keys
+pub type Keyring = Vec<SignedPublicKey>;
+
+/// Create a Keyring from a dir of public keys
+pub fn load_public_keys() -> Fallible<Keyring> {
+  let mut result: Keyring = vec![];
+  for entry in read_dir(PUBKEYS_DIR).context("Reading public keys dir")? {
+    let path = &entry?.path();
+    let path_str = match path.to_str() {
+      None => continue,
+      Some(p) => p,
+    };
+    let file = File::open(path).context(format!("Reading {}", path_str))?;
+    let (pubkey, _) =
+      SignedPublicKey::from_armor_single(file).context(format!("Parsing {}", path_str))?;
+    match pubkey.verify() {
+      Err(err) => return Err(format_err!("{:?}", err)),
+      Ok(_) => result.push(pubkey),
+    };
+  }
+  Ok(result)
+}
+
+/// Verify that signature is valid and contains expected digest
+pub async fn verify_signature(
+  public_keys: &Keyring,
+  body: Bytes,
+  expected_digest: &str,
+) -> Fallible<()> {
+  let msg = Message::from_bytes(body.reader()).context("Parsing message")?;
+
+  // Verify signature using provided public keys
+  if !public_keys.iter().any(|ref k| msg.verify(k).is_ok()) {
+    return Err(format_err!("No matching key found to decrypt {:#?}", msg));
```

Comment:
s/decrypt/verify/?

Also, we can extract the claimed signer's key ID [from the message][1], so we shouldn't need to iterate over trusted keys; we can just verify vs. that claimed key (or complain that the claimed key ID is unrecognized).  But iterating over a handful of keys should be cheap enough, if the PGP library doesn't expose the issuer conveniently.

[1]: https://tools.ietf.org/html/rfc4880#section-5.2.3.5

Reply:
Ah, good idea, seems we can fetch KeyID via https://docs.rs/pgp/0.7.1/pgp/composed/message/enum.Message.html#method.get_recipients

Reply:
Hmm, it returns empty and `key_id` is private is most signature types.

I'd rather leave it for later